gravatar/src/main/java/com/gravatar/services/ProfileService.kt

```diff
@@ -79,7 +81,67 @@ public class ProfileService(private val okHttpClient: OkHttpClient? = null) {
      * @param username The username to fetch the profile for
      * @param getProfileListener The listener to notify of the fetch result
      */
-    public fun fetchByUsername(username: String, getProfileListener: GravatarListener<UserProfiles>) {
-        fetch(username, getProfileListener = getProfileListener)
+    public fun fetchWithListenerByUsername(username: String, getProfileListener: GravatarListener<UserProfiles>) {
+        fetchWithListener(username, getProfileListener = getProfileListener)
+    }
+
+    /**
+     * Exception thrown when a fetch operation fails.
+     *
+     * @property errorType The type of error that occurred
+     */
+    public class FetchException(public val errorType: ErrorType) : Exception()
+
+    /**
+     * Fetches a Gravatar profile for the given hash or username.
+     *
+     * @param hashOrUsername The hash or username to fetch the profile for
+     * @return The fetched profile
+     */
+    public suspend fun fetch(hashOrUsername: String): UserProfiles {
+        return suspendCoroutine {
+            fetchWithListener(
+                hashOrUsername,
+                object : GravatarListener<UserProfiles> {
+                    override fun onSuccess(response: UserProfiles) {
+                        it.resumeWith(Result.success(response))
+                    }
+
+                    override fun onError(errorType: ErrorType) {
+                        it.resumeWithException(FetchException(errorType))
```

Comment:
❓Have you considered returning the error instead of throwing an exception? Using a Sealed class, for example. Throwing an exception can crash the app if you don't have a try/catch surrounding it. As far as I've read, there is no way in Kotlin to force the exception to be caught, so I wonder if we should use another approach.

Reply:
I don't have strong opinion on this. I read code from folks using a sealed class that embeds the result or the error under a Success/Failure subclass. But they need to pack and unpack the error which results in another abstraction. 

With exceptions, we could just drop the `handleError` call and forward the throwable that was created in the underlying layer instead of packing them into an ErrorType, eg. we could do something like:

```kotlin
try {
    profile = fetch()
} catch (GravatarException e) {
    // specific to gravatar (eg. unknown hash?)
    showError()
} catch (SocketTimeoutException e) {
    // recoverable network error, show a retry button or auto retry later on connectivity change
    showErrorAndRetryButton()
}
```

Reply:
If I'm not mistaken, same logic with a sealed class code will look like:

```kotlin
var error = null
var profile = null
val result = fetch()
when (result) {
    is Success -> profile = result.data
    is Error -> error = result.error
}
when (error) {
    HASH_NOT_FOUND -> showError()
    TIMEOUT -> showErrorAndRetryButton()
}
```

Reply:
>  as a third-party developer, if a call profile = fetch(), I won't have any warning from the IDE when using an exception, 

Fair enough, I'll drop the exception and replace it with a `Result` class

Reply:
Updated in [ff53990](https://github.com/Automattic/Gravatar-SDK-Android/pull/130/commits/ff539903db402693f97dd023078d772b5ea35490)